File: problems/0079/solution.cpp
#include <vector>
using namespace std;

class Solution {
 private:
  int dirs[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
  void gridDFS(vector<vector<char>>& board, int r, int c, string word,
               int currentIndex, bool& res) {
    if (r < 0 || r >= board.size() || c < 0 || c >= board[0].size() ||
        board[r][c] == '.' || res) {
      return;
    }

    char tmp = board[r][c];
    if (tmp == word[currentIndex]) {
      if (currentIndex == word.size() - 1) {
        res = true;
        return;
      }
      board[r][c] = '.';
      for (auto& d : dirs) {
        gridDFS(board, r + d[0], c + d[1], word, currentIndex + 1, res);
      }
      board[r][c] = tmp;
    }
  }

 public:
  bool exist(vector<vector<char>>& board, string word) {
    bool res = false;
    for (int i = 0; i < board.size(); i++) {
      for (int j = 0; j < board[0].size(); j++) {
        if (board[i][j] == word[0]) {
          gridDFS(board, i, j, word, 0, res);
        }
        if (res) {
          return res;
        }
      }
    }
    return res;
  }
};

File: problems/0079/solution.py
from typing import List


class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        res: bool = False

        def grid_dfs(r: int, c: int, current_index: int):
            nonlocal res
            if (
                r < 0
                or r >= len(board)
                or c < 0
                or c >= len(board[0])
                or board == "."
                or res
            ):
                return
            dirs = [(-1, 0), (1, 0), (0, -1), (0, 1)]
            tmp = board[r][c]
            if tmp == word[current_index]:
                if current_index == len(word) - 1:
                    res = True
                    return
                board[r][c] = "."
                for dr, dc in dirs:
                    grid_dfs(r + dr, c + dc, current_index + 1)
                board[r][c] = tmp

        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == word[0]:
                    grid_dfs(i, j, 0)
                if res:
                    return res
        return res
